### PR TITLE
Benchmarking V2: framework impl

### PR DESCRIPTION
**Pull request by ahadnagy**
# What does this PR do?

This PR is another iteration of reworking the benchmarking flow in Transformers. the goal is to have a similar flow as for Diffusers: daily reports to HF Datasets, and visualization in Gradio Spaces.

This PR focuses on the framework fundamentals, export to Datasets, GH actions and more model support will come in follow-ups.

From the wishlist, the first iteration includes the following
- JSON output with all the scenarios as an array
- support for different attention packages and SDPA backends
- compiled, kernelized scenarios
- HW info an utilization collection
- abstractions for making the ModelBenchmark code more lean and standardized.

I put everything into a _v2 folder so we can keep the existing framework intact until this stabilizes a bit.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @ydshieh maybe? not sure who to ping for daily benchmark runs

**Comment by ydshieh** [comment]
@McPatate would be better person :-)

File: benchmark_v2/README.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+```bash
+# Run all benchmarks with default settings
+python run_benchmarks.py
+
+# Specify output directory
+python run_benchmarks.py --output-dir my_results
+
+# Run with custom parameters
+python run_benchmarks.py \
+    --warmup-iterations 5 \
```
Comment (McPatate) [inline benchmark_v2/README.md:19]: Nice to have this as a flag!

File: benchmark_v2/README.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+          "gpu_memory_total": 40960,
+          "cpu_count": 64
+        },
+        "config": {
+          "variant": "eager",
+          "warmup_iterations": 3,
+          "measurement_iterations": 5
+        }
+      },
+      "measurements": {
```
Comment (McPatate) [inline benchmark_v2/README.md:57]: Would it make sense to have individual events (or aggregates over small time buckets) to be able to plot the data on top of having variance / means?

Comment (McPatate) [reply benchmark_v2/README.md:57]: Maybe in later version anyway

File: benchmark_v2/run_benchmarks.py [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+    numeric_level = getattr(logging, log_level.upper(), None)
+    if not isinstance(numeric_level, int):
+        raise ValueError(f'Invalid log level: {log_level}')
+    
+    logging.basicConfig(
+        level=numeric_level,
+        format='[%(levelname)s - %(asctime)s] %(name)s: %(message)s',
+        handlers=[
+            logging.StreamHandler(sys.stdout),
+            logging.FileHandler(f'benchmark_run_{datetime.now().strftime("%Y%m%d_%H%M%S")}.log')
```
Comment (McPatate) [inline benchmark_v2/run_benchmarks.py]: Not sure we really need a file logger, maybe enable this with a cmd line arg?

File: benchmark_v2/run_benchmarks.py [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+        return None
+
+
+def generate_summary_report(
+    output_dir: str, 
+    benchmark_results: Dict[str, Any],
+    logger: logging.Logger
+) -> str:
+    """Generate a summary report of all benchmark runs."""
+    summary_file = os.path.join(output_dir, "benchmark_summary.json")
```
Comment (McPatate) [inline benchmark_v2/run_benchmarks.py]: Perhaps we should add a timestamp to the file name so multiple runs won't overwrite the existing content.
Also, not sure a file is needed, stdout is ok imo

**Review by McPatate — COMMENTED** [review]
Did a first pass, will continue tomorrow.
Awesome work! 🔥

File: benchmark_v2/benches/llama.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+import os
+import logging
+from typing import Dict, Any, List
+
+from benchmark_framework import ModelBenchmark
+
+import torch
+
+os.environ["HF_HUB_ENABLE_HF_TRANSFER"] = "1"
```
Comment (McPatate) [inline benchmark_v2/benches/llama.py:23]: Is this necessary now that we use xet for most repos?

File: benchmark_v2/benchmark_framework.py [hunk 6]
```diff
@@ -0,0 +1,1181 @@
...
+import logging
+
+import numpy as np
+import psutil
+import gpustat
+
+import torch
+
+
+class CUDATimer:
```
Comment (McPatate) [inline benchmark_v2/benchmark_framework.py]:
```suggestion
class GPUTimer:
```
perhaps? `CUDATimer` sounds a bit narrow given cuda is optional

Comment (McPatate) [reply benchmark_v2/benchmark_framework.py]: Or something like `ArchAwareTimer` idk 😄

File: benchmark_v2/benchmark_framework.py [hunk 7]
```diff
@@ -0,0 +1,1181 @@
...
+        return f"BenchmarkScenario(name='{self.name}', variant='{self.config.variant}')"
+
+
+
+
+@dataclass
+class TimingResult:
+    """Result from a timing measurement."""
+    time_to_first_token: Optional[float] = None
+    latency: float = 0.0
```
Comment (McPatate) [inline benchmark_v2/benchmark_framework.py]: I think it could be neat to have the unit in the name like so:
```suggestion
    time_to_first_token_seconds: Optional[float] = None
    latency_seconds: float = 0.0
```
for clarity, like below

Comment (McPatate) [reply benchmark_v2/benchmark_framework.py]: or at least a doc string for each parameter

File: benchmark_v2/benchmark_framework.py [hunk 8]
```diff
@@ -0,0 +1,1181 @@
...
+    tokens_per_second: Optional[float] = None
+    time_per_output_token_seconds: Optional[float] = None
```
Comment (McPatate) [inline benchmark_v2/benchmark_framework.py:206]: Isn't this the same as latency?

File: benchmark_v2/benchmark_framework.py [hunk 9]
```diff
@@ -0,0 +1,1181 @@
...
+                    logger.warning(f"Teardown callback failed for scenario {self.name}: {e}")
+    
+    def __repr__(self):
+        return f"BenchmarkScenario(name='{self.name}', variant='{self.config.variant}')"
+
+
+
+
+@dataclass
+class TimingResult:
```
Comment (McPatate) [inline benchmark_v2/benchmark_framework.py:201]: Is this for one event or is this storing averages?

File: benchmark_v2/benchmark_framework.py [hunk 10]
```diff
@@ -0,0 +1,1181 @@
...
+            p99=float(np.percentile(measurements_array, 99)),
+            unit=unit
+        )
+
+
+@dataclass 
+class HardwareInfo:
+    """Hardware information collected during benchmarking."""
+    gpu_name: str
+    gpu_memory_total: int  # MB
```
Comment (McPatate) [inline benchmark_v2/benchmark_framework.py]: To be coherent with the other variables, I'd add unit in the name of the variable, or rename the other variables and just add docs strings describing unit.
But given that you are creating the JSON from dataclasses I assume it makes sense to have the unit in the variable's name imo
```suggestion
    gpu_memory_total_mb: int
```

File: benchmark_v2/benchmark_framework.py [hunk 11]
```diff
@@ -0,0 +1,1181 @@
...
+    cpu_count: int
+    memory_total: int  # MB
```
Comment (McPatate) [inline benchmark_v2/benchmark_framework.py]: same here
```suggestion
    memory_total_mb: int
```

File: benchmark_v2/benchmark_framework.py [hunk 12]
```diff
@@ -0,0 +1,1181 @@
...
+            
+        self.monitoring = True
+        self.gpu_utilization = []
+        self.gpu_memory_used = []
+        self.timestamps = []
+        self.thread = threading.Thread(target=self._monitor_loop)
+        self.thread.start()
+        self.logger.debug("GPU monitoring started")
+        
+    def stop(self):
```
Comment (McPatate) [inline benchmark_v2/benchmark_framework.py]:
```suggestion
    def stop_and_collect(self):
```

Comment (McPatate) [reply benchmark_v2/benchmark_framework.py]: Could also be nice to have a type for the return, even if it's a `Union[WithGPU, NoGPU]`

File: benchmark_v2/benchmark_framework.py [hunk 13]
```diff
@@ -0,0 +1,1181 @@
...
+                    gpu = gpu_stats[0]
+                    self.gpu_utilization.append(gpu["utilization.gpu"])
+                    self.gpu_memory_used.append(gpu["memory.used"])
+                    self.timestamps.append(time.time())
+                    consecutive_failures = 0  # Reset failure counter on success
+                else:
+                    consecutive_failures += 1
+                    if consecutive_failures >= max_consecutive_failures and not self.error_logged:
+                        self.logger.warning("GPU monitoring: No GPU data returned by gpustat")
+                        self.error_logged = True
```
Comment (McPatate) [inline benchmark_v2/benchmark_framework.py]:
```suggestion
                        self.warning_logged = True
```

File: benchmark_v2/benchmark_framework.py [hunk 14]
```diff
@@ -0,0 +1,1181 @@
...
+    config: BenchmarkConfig
+
+
+class GPUMonitor:
+    """Monitor GPU utilization during benchmark execution."""
+    
+    def __init__(self, sample_interval: float = 0.1, logger: logging.Logger = None):
+        self.sample_interval = sample_interval
+        self.logger = logger or logging.getLogger(__name__)
+        self.monitoring = False
```
Comment (McPatate) [inline benchmark_v2/benchmark_framework.py]: Is using an [`Event`](https://docs.python.org/3/library/threading.html#event-objects) not preferable here?

File: benchmark_v2/benchmark_framework.py [hunk 15]
```diff
@@ -0,0 +1,1181 @@
...
+        self.config = None
+        self._default_prompt = "Why dogs are so cute?"
+        
+    @property
+    def default_prompt(self) -> str:
+        """Default prompt for text generation. Override in subclasses if needed."""
+        return self._default_prompt
+    
+    @property
+    def model_type(self) -> str:
```
Comment (McPatate) [inline benchmark_v2/benchmark_framework.py]: What's the purpose of `model_type`?

File: benchmark_v2/benchmark_framework.py [hunk 16]
```diff
@@ -0,0 +1,1181 @@
...
+                    "scenario_description": scenario.description
+                }
+                
+                if latency_measurements:
+                    latency_stats = BenchmarkStatistics.from_measurements("latency", latency_measurements)
+                    scenario_results["measurements"]["latency"] = asdict(latency_stats)
+                
+                if ttft_measurements:
+                    ttft_stats = BenchmarkStatistics.from_measurements("time_to_first_token", ttft_measurements)
+                    scenario_results["measurements"]["time_to_first_token"] = asdict(ttft_stats)
```
Comment (McPatate) [inline benchmark_v2/benchmark_framework.py]:
```suggestion
                    scenario_results["measurements"]["time_to_first_token_seconds"] = asdict(ttft_stats)
```

File: benchmark_v2/benchmark_framework.py [hunk 17]
```diff
@@ -0,0 +1,1181 @@
...
+                
+                # Log summary
+                if latency_measurements:
+                    self.logger.info(f"Latency: {latency_stats.mean:.4f}±{latency_stats.std:.4f}s (mean±std)")
+                if ttft_measurements:
+                    self.logger.info(f"TTFT: {ttft_stats.mean:.4f}±{ttft_stats.std:.4f}s (mean±std)")
+                if tokens_per_sec_measurements:
+                    self.logger.info(f"Throughput: {tps_stats.mean:.2f}±{tps_stats.std:.2f} tokens/sec (mean±std)")
+                if tpot_measurements:
+                    self.logger.info(f"TPOT: {tpot_stats.mean:.4f}±{tpot_stats.std:.4f}s/token (mean±std)")
```
Comment (McPatate) [inline benchmark_v2/benchmark_framework.py]: Is `TPOT` clear/widely used?

File: benchmark_v2/benches/llama.py [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+            include_sdpa_variants=include_sdpa_variants,
+            device=device,
+            torch_dtype=torch_dtype,
+            batch_size=batch_size
+        )
+        
+        logger.info(f"Created {len(scenarios)} benchmark scenarios")
+        
+        # Create runner and execute benchmarks
+        runner = BenchmarkRunner(logger, output_dir)
```
Comment (McPatate) [inline benchmark_v2/benches/llama.py:138]: Will there be cases where we run multiple `Benchmark`s with a single `BenchmarkRunner`?

File: benchmark_v2/benchmark_framework.py (hunk 8 above)
Comment (McPatate) [reply benchmark_v2/benchmark_framework.py:206]: Ok just [googled](https://bentoml.com/llm/inference-optimization/llm-inference-metrics) what everything is, I think ITL (inter-token latency) is clearer than TPOT, but from what I can see it looks like TPOT is more widely used.
Do you think it's ok for us to use ITL instead?

File: benchmark_v2/benches/llama.py (hunk 5 above)
Comment (ahadnagy) [reply benchmark_v2/benches/llama.py:23]: Not really, removing it.

File: benchmark_v2/run_benchmarks.py (hunk 3 above)
Comment (ahadnagy) [reply benchmark_v2/run_benchmarks.py]: It proved itself useful for development, I'll put it behind a cmd flag - disabled by default.

File: benchmark_v2/run_benchmarks.py (hunk 4 above)
Comment (ahadnagy) [reply benchmark_v2/run_benchmarks.py]: On a second thought, I'm not even sure this file is useful.

File: benchmark_v2/benchmark_framework.py (hunk 8 above)
Comment (ahadnagy) [reply benchmark_v2/benchmark_framework.py:206]: I'm fine with ITL. Personally, I'm a tok/sec guy, but we also have that. :)

File: benchmark_v2/README.md (hunk 2 above)
Comment (ahadnagy) [reply benchmark_v2/README.md:57]: I'm all for doing this in a later version. This is fine as a first step, but we can definitely stat nerd the thing.

File: benchmark_v2/benchmark_framework.py (hunk 9 above)
Comment (ahadnagy) [reply benchmark_v2/benchmark_framework.py:201]: This is for storing the results of a complete benchmark scenario. This is the data that gets serialized into the JSON output. (and this is where we could add the time-series data later)

File: benchmark_v2/benchmark_framework.py (hunk 15 above)
Comment (ahadnagy) [reply benchmark_v2/benchmark_framework.py]: Leftover from another iteration, removed it. :)

File: benchmark_v2/benches/llama.py (hunk 18 above)
Comment (ahadnagy) [reply benchmark_v2/benches/llama.py:138]: `BenchmarkRunner` will always run the benchmark for a single model, but all the scenarios for that model.

File: benchmark_v2/benchmark_framework.py [hunk 19]
```diff
@@ -0,0 +1,1204 @@
...
+import logging
+
+import numpy as np
+import psutil
+import gpustat
+
+import torch
+
+
+class WithGPU(TypedDict):
```
Comment (McPatate) [inline benchmark_v2/benchmark_framework.py]: I didn't necessarily mean to take my suggestion literally, it was the essence of it that was important 😄 

```suggestion
class GPUMetrics(TypedDict):
```
may be better for this one

File: benchmark_v2/benchmark_framework.py [hunk 20]
```diff
@@ -0,0 +1,1204 @@
...
+    gpu_utilization_max: float
+    gpu_utilization_min: float
+    gpu_memory_used_mean: float
+    gpu_memory_used_max: float
+    gpu_memory_used_min: float
+    sample_count: int
+    gpu_monitoring_status: str
+
+
+class NoGPU(TypedDict):
```
Comment (McPatate) [inline benchmark_v2/benchmark_framework.py:49]: That one I don't know, this should be good enough

File: benchmark_v2/benchmark_framework.py (hunk 9 above)
Comment (McPatate) [reply benchmark_v2/benchmark_framework.py:201]: In that case perhaps we should add `_mean[_]` to field names of this class? I assume the values are means.

Comment (ahadnagy) [reply benchmark_v2/benchmark_framework.py:201]: Pardon, late evening brainfart. `TimingResult` represents the data point of a benchmark iteration, while `BenchmarkStatistics` holds the derived stats.

File: benchmark_v2/benchmark_framework.py (hunk 19 above)
Comment (ahadnagy) [reply benchmark_v2/benchmark_framework.py]: GPUMetrics sounds a bit better, yes. Done.

File: benchmark_v2/benchmark_framework.py (hunk 20 above)
Comment (ahadnagy) [reply benchmark_v2/benchmark_framework.py:49]: Kept this one, seems to capture the essence well. :D

File: benchmark_v2/run_benchmarks.py (hunk 4 above)
Comment (ahadnagy) [reply benchmark_v2/run_benchmarks.py]: I kept it at the end with a timestamp.

**Review by McPatate — APPROVED** [review]
lgtm overall!

Do you have the space ready? 👀

**Comment by ahadnagy** [comment]
Thank you! The space is coming with tomorrows's PR and GH Actions setup, will post it here as well :)
